### DISCUSSION TITLE
Computing KKT conditions of IPOPT, and sensitivity with respect to parameters for asNMPC

### DISCUSSION THREAD

**Discussion by MarkusSvedberg** [casadi/casadi#3156] [Q&A]
Hello!

I am considering implementing asNMPC for solving a receeding horizon OCP, based on this [paper](https://www.sciencedirect.com/science/article/pii/S0005109808004196?via%3Dihub)

I previously found this [discussion](https://groups.google.com/g/casadi-users/c/plTu2j-Anwg/m/1xSUDWAGIgAJ), about it not being feasible with IPOPT yet (jan 2022). I was looking to see if the sensitivity computations have matured at this point? 

I did some attempts by myself, and I believe i computed the KKT conditions by doing the following (python)

```python
...
self.NLP = {'f': self.cost, 'x': self.decision_variables, 'g': self.constraints, 'p': self.parameters}
self.solver = ca.nlpsol('solver', 'ipopt', self.NLP, options)
...
phi = self.solver.jacobian()
```

where I think $\varphi=$`phi` is the KKT conditions? Please correct me if I am wrong here. If not, how would I go about calculating 

$$\frac{\partial\varphi}{\partial s}$$

where $s$ is a set containing decision variables, and lagrange multipliers? I can't seem to be able to call

```python
ca.jacobian(phi, s)
```

since phi is a function. And

```python
phi.jacobian()
```

is not callable with any arguments, and calling it as is I believe calculates the entire jacobian, which obviously is needlessly expensive (I ran out of memory). 

If knows what to do, or has any ideas, please let me know!

Best,
Markus Svedberg

**Comment by jaeandersson** [comment]
> I previously found this [discussion](https://groups.google.com/g/casadi-users/c/plTu2j-Anwg/m/1xSUDWAGIgAJ), about it not being feasible with IPOPT yet (jan 2022). I was looking to see if the sensitivity computations have matured at this point?

I think the issue with IPOPT has always been that it doesn't deliver the exact active set and the CasADi NLP sensitivity analysis is based on knowledge of the active set. In Biegler's group they have implemented sensitivity analysis approaches that work differently - I think the first effort was the sIPOPT package. But no such alternative sensitivity approach has been interfaced to CasADi and there are no plans to do so.

What is maturing (slowly) is CasADi's native NLP sensitivity analysis support, based on applying the implicit function theorem to the KKT conditions. That is implemented in the NLP base classes so it applies to all solvers. You could still use IPOPT, in principle, to get close to the solution. And then initialize one of the native CasADi NLP solvers at that point as shown in examples.

**Reply by brunoml5** [reply]
Hi!

1. Is there an alternative toolbox that implements NLP sensitivity analysis with more mature algorithms?
2. Is asNMPC implemented in some toolbox? Maybe some toolbox from Biegler's group

I'm just curious, it is not important.

**Reply by MarkusSvedberg** [reply]
Thank you for your quick answer! 

> You could still use IPOPT, in principle, to get close to the solution. And then initialize one of the native CasADi NLP solvers at that point as shown in examples.

I saw those examples, but were not sure how to apply them here, as they use opti, and I have not managed to translate that code to the `nlpsol` regime. I am trying to replicate the following from [here](https://web.casadi.org/blog/nlp_sens/)

```matlab
opts.qpsol = 'qrqp';
opti.solver('sqpmethod',opts);
Z = opti.to_function('Z',{p,xy},{z(xy)});
zp = Z(p,xy);
``` 

What I have done is 

```python
solver = ca.nlpsol('solver', 'sqpmethod', self.NLP, dict(qpsol='qrqp'))
sol_analytic = solver(x0=sol["x"], p=params_analytic, lbx=lbx, ubx=ubx, lbg=lbg, ubg=ubg)
``` 

where `sol` is the solution found by IPOPT and `params_analytic` is the same parameter vector as used when solving with IPOPT, but with one of the parameters defined as an SX instead. I assume I am doing something wrong when calling the sovler, but I am not sure how to make the call "symbolically". I get the error

```
'eval_sx' not defined for Sqpmethod
``` 

Any help would be appeciated!

Best,
Markus

**Reply by jaeandersson** [reply]
To be clear, I think the NLP sensitivity approach in CasADi is clean and it has been shown to work well in multiple applications. For challenging applications, you might need some support from CasADi developers to get it to work efficiently, but that's the case with a lot of advanced features of CasADi. The main challenge as I said is getting the right active set which is more about having access to the right NLP and QP solvers than the sensitivity approach per se.

Concerning asNMPC, my understanding is that it's pretty similar to the kind of fast-feedback algorithms you see implemented in tools like ACADOS, so that could be a place to start. Victor Zavala these days is mostly programming in Julia, so I wouldn't be surprised if you can find something written in Julia too.

**Reply by FreyJo** [reply]
>     Is asNMPC implemented in some toolbox? Maybe some toolbox from Biegler's group

`acados` implements the Advanced-Step Real-Time Iterations (AS-RTI) which can be seen as a generalization of asNMPC.
This work is described in the following publication:

> Advanced-Step Real-Time Iterations With Four Levels – New Error Bounds and Fast Implementation in acados, J. Frey, A. Nurkanović, M. Diehl, IEEE Control Systems Letters 2024

Hope you find it useful :)

**Comment by yakunouyang** [comment]
> To be clear, I think the NLP sensitivity approach in CasADi is clean and it has been shown to work well in multiple applications. For challenging applications, you might need some support from CasADi developers to get it to work efficiently, but that's the case with a lot of advanced features of CasADi. The main challenge as I said is getting the right active set which is more about having access to the right NLP and QP solvers than the sensitivity approach per se.

According to what I found in #3131, the active set output from IPOPT can also be used by NLP sensitivity approach in CasADi as long as the following parameters are set:

```
clip_inactive_lam: true
sens_linsol: ma27
```

A working example: 

```python
from casadi import *

x = MX.sym("x", 2)
p = MX.sym("p")

f = (x[0]-1)**2 + (x[1]+1)**2
g = x - p

options = {"clip_inactive_lam": True, "sens_linsol": 'ma27'}
solver = nlpsol('solver', 'ipopt', {"x": x, "p": p, "f": f, "g": g}, options)

f = Function('f', [p], [solver(p=p, lbx=-10, ubx=10, lbg=0, ubg=inf)["x"]])
J = Function('J', [p], [f(p), jacobian(f(p), p)])

[sol, jac_x_p] = J(2.0)
print('solution: ', sol)
print('dx/dp: ', jac_x_p)
```

**Reply by yakunouyang** [reply]
related issues: #2545 #2660

**Reply by jaeandersson** [reply]
> > To be clear, I think the NLP sensitivity approach in CasADi is clean and it has been shown to work well in multiple applications. For challenging applications, you might need some support from CasADi developers to get it to work efficiently, but that's the case with a lot of advanced features of CasADi. The main challenge as I said is getting the right active set which is more about having access to the right NLP and QP solvers than the sensitivity approach per se.
> 
> According to what I found in #3131, the active set output from IPOPT can also be used by NLP sensitivity approach in CasADi as long as the following parameters are set:
> 
> ```
> clip_inactive_lam: true
> sens_linsol: ma27
> ```
> 
> A working example:
> 
> ```python
> from casadi import *
> 
> x = MX.sym("x", 2)
> p = MX.sym("p")
> 
> f = (x[0]-1)**2 + (x[1]+1)**2
> g = x - p
> 
> options = {"clip_inactive_lam": True, "sens_linsol": 'ma27'}
> solver = nlpsol('solver', 'ipopt', {"x": x, "p": p, "f": f, "g": g}, options)
> 
> f = Function('f', [p], [solver(p=p, lbx=-10, ubx=10, lbg=0, ubg=inf)["x"]])
> J = Function('J', [p], [f(p), jacobian(f(p), p)])
> 
> [sol, jac_x_p] = J(2.0)
> print('solution: ', sol)
> print('dx/dp: ', jac_x_p)
> ```

I think what can happen is that some constraints are active, but pretty weakly active. Clipping in that case might not work because the KKT system can become singular. Often, you want to calculate sensitivities exactly for that type of problems, to see which confidence intervals are very large, which could inform how to best reformulate the problem. In that case, you might be better off using an active set QP solver initialized at the ipopt solution.

**Reply by sandeep026** [reply]
@MarkusSvedberg [csnlp](https://casadi-nlp.readthedocs.io/en/latest/auto_examples/sensitivity/comparison_sensitivities.html#sphx-glr-auto-examples-sensitivity-comparison-sensitivities-py). This might be of interest.

**Reply by FreyJo** [reply]
Also note solution sensitivity computations have been implemented in `acados`.
They are especially efficient when considering OCPs, but examples with dense NLPs can be formulated with horizon 0.
See our publication on "Differentiable Nonlinear Model Predictive Control" -- https://arxiv.org/abs/2505.01353
